gadget/gadget.go

```diff
@@ -551,6 +563,117 @@ type VolumeUpdate struct {
+	if len(va.Assignments) == 0 {
+		return fmt.Errorf("no assignments specified")
+	}
+
+	for name, asv := range va.Assignments {
+		if v := volumes[name]; v == nil {
+			return fmt.Errorf("volume %q is mentioned in assignment but has not been defined", name)
+		}
+		if err := asv.validate(); err != nil {
+			return fmt.Errorf("%q: %v", name, err)
+		}
+	}
+	return nil
+}
+
+// DeviceAssignment is the device data for each volume assignment. Currently
+// just keeps the device the volume should be assigned to.
+type DeviceAssignment struct {
+	// Device is the actual block device available in the system,
+	// eg. /dev/disk/by-path/pci-42:0
+	Device string `yaml:"device"`
+}
+
+func (da *DeviceAssignment) validate() error {
+	// Device disk paths must be either under by-path or by-id
+	// as those are the only sensible ones that can be predefined
+	if !(strings.HasPrefix(da.Device, "/dev/disk/by-path") || strings.HasPrefix(da.Device, "/dev/disk/by-id")) {
+		return fmt.Errorf("unsupported device path %q, for now only paths under /dev/disk/{by-path,by-id} are valid", da.Device)
+	}
+	return nil
+}
+
+func areAssignmentsMatchingCurrentDevice(name string, assignments map[string]*DeviceAssignment) bool {
+	for _, va := range assignments {
+		if _, err := os.Stat(path.Join(dirs.GlobalRootDir, va.Device)); err != nil {
+			// XXX: expect this to mean no path, consider actually
+			// ensuring the error is not-exists
+			logger.Debugf("cannot stat device %s for volume assignment %q: %v",
+				va.Device, name, err)
+			return false
+		}
+	}
+	return true
+}
+
+// Indirection here for mocking
+var VolumesForCurrentDeviceAssignment = volumesForCurrentDeviceAssignmentImpl
+
+// volumesForCurrentDeviceAssignmentImpl finds a matching assignment based on
+// the available /dev/disk environment. If none, or multiple matching are found,
+// an error is returned.
+func volumesForCurrentDeviceAssignmentImpl(gi *Info) (map[string]*Volume, error) {
+	var matched []*VolumeAssignment
+	for _, vas := range gi.VolumeAssignments {
+		if !areAssignmentsMatchingCurrentDevice(vas.Name, vas.Assignments) {
+			continue
+		}
+		matched = append(matched, vas)
+	}
+
+	// If we find nothing, throw an error about no assignments being matched
+	// for the current device. If we find more than one, error out on that too.
+	switch {
+	case len(matched) == 0:
+		return nil, fmt.Errorf("no matching volume-assignment for current device")
+	case len(matched) == 1:
+		// update volume assignments, we can assume only one match
+		// in the list.
+		vas := matched[0]
+		logger.Noticef("found valid device-assignment: %s", vas.Name)
+		volumes := make(map[string]*Volume)
+		for vol := range vas.Assignments {
+			gi.Volumes[vol].AssignedDevice = vas.Assignments[vol].Device
+			volumes[vol] = gi.Volumes[vol]
+		}
+		return volumes, nil
+	default:
+		return nil, fmt.Errorf("multiple matching volume-assignment for current device")
```

Comment:
Such use case feel very finicky and is basically asking for trouble in the long run. If we do this it would require visibility to make sure that the thing that matched by accident is what was intended.

Reply:
There are other cases I can think about, for instance:
- vendor creates a SKU with one disk, deploys it around
- vendor makes a modification of the design and adds a disk. They need then different assignments for the 2 SKUs and want to actually use the same gadget/image. The first SKU would match the 2 possible assignements, so not allowing that would produce an error.

Again, I think that the intention of the spec is to allow this sort of scenario, it is not a match by accident, it is that I do not see crazy that there is some intersection between assignements and to me what spec is saying is that because of that we need to use the order in the list as priority.

@kubiko in your experience in the field, does this make sense?

Reply:
Hmm tricky one @alfonsosanchezbeato 
It is common to have multiple sources for components like eMMC. So if we have descriptive laguage which pins it down as `/dev/disk/by-path/pci-0000:02:00.1-ata-5` then having multiple options woud make sense.
But.... Otherside of the coin is that ` /dev/disk/by-path/xxxxx` in still symbolic link to something which is more predictable e.g. `/dev/mmcblk0p1`. And this needs to remain the same for all the different sources. Otherwise when mmc device enumeration changes, u-boot mmc device number would change as well and we would need new u-boot build per SKU. (sure one can add some abiguity to u-boot scrip, but that on it's own is wrong approach)

So I see @alfonsosanchezbeato point, but we are potentially overthinking it. IMHO in practive ppl will use `/dev/mmcblk0p1` rather than `/dev/disk/by-path/`
I checked also device with UFS storage, and there we know it's `/dev/sda`, `/dev/sdb`,.....
Also alternatice sources would be still plugged to the same interface, so they would likely still show as the same `by-path` anyway.

Reply:
The spec is actually not super clear on this point, the intention is that there we pick the first matching assignments (all nodes can be resolved), that is true, but is not explicit about ambiguity. I think for now if there are many matching assignments in theory we should produce an error only if they really result in different assignments, that means really ambiguity, this assumes we can find out if different device paths actually point to the same device?

Reply:
@kubiko note that here we allow only `by-path` and `by-id` as those identifiers are the only ones that could not change across reboots if there are disks of the same type.

@pedronis that's a good point, but taking into account that we use only `by-path` and `by-id`, in practice it would happen only if there is an assignement using `by-id` and a different one using `by-path` that would point to the same device, which would be strange as you would not want to define one using `by-id` in that case possibly.

Anyway, I am fine with current approach if there is no traction in what I suggest, as it is more restrictive so it can be made more flexible later.